src/auth/oauth/errors.rs

```diff
@@ -0,0 +1,167 @@
+use super::ValidatedRedirectUri;
+use crate::auth::AuthenticationError;
+use crate::models::error::ApiError;
+use crate::models::ids::DecodingError;
+use actix_web::http::StatusCode;
+use actix_web::HttpResponse;
+
+#[derive(thiserror::Error, Debug)]
+#[error("{}", .error_type)]
+pub struct OAuthError {
+    #[source]
+    pub error_type: OAuthErrorType,
+
+    pub state: Option<String>,
+    pub valid_redirect_uri: Option<ValidatedRedirectUri>,
+}
+
+impl OAuthError {
+    /// The OAuth request failed either because of an invalid redirection URI
+    /// or before we could validate the one we were given, so return an error
+    /// directly to the caller
+    ///
+    /// See: IETF RFC 6749 4.1.2.1 (https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.2.1)
+    pub fn error(error_type: impl Into<OAuthErrorType>) -> Self {
+        Self {
+            error_type: error_type.into(),
+            valid_redirect_uri: None,
+            state: None,
+        }
+    }
+
+    /// The OAuth request failed for a reason other than an invalid redirection URI
+    /// So send the error in url-encoded form to the redirect URI
+    ///
+    /// See: IETF RFC 6749 4.1.2.1 (https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.2.1)
+    pub fn redirect(
+        err: impl Into<OAuthErrorType>,
+        state: &Option<String>,
+        valid_redirect_uri: &ValidatedRedirectUri,
+    ) -> Self {
+        Self {
+            error_type: err.into(),
+            state: state.clone(),
+            valid_redirect_uri: Some(valid_redirect_uri.clone()),
+        }
+    }
+}
+
```

Comment:
This is to adhere to the OAuth spec as outlined in sections 4.1.2.1

Reply:
This is to adhere to oauth spec as outlined in [sections 4.1.2.1](https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.2.1) and [5.2](https://datatracker.ietf.org/doc/html/rfc6749#section-5.2).

Basically, instead of doing what we do for ApiError by returning error code + json-serialized error description, we need to do one of two things:
1. If the redirect uri is valid, then redirect the user to the callback location with error information provided as query parameters
2. If the redirect uri is invalid, generally, then we return an error directly
As we do these, to adhere to the spec, we need to use the codes and terms as outlined in RFC 6749.

Given this amount of separate behavior and the importance of isolating it from change that may come to ApiError that could bring it off-spec, it made sense to give it its own error type.

Reply:
This makes sense- as oauth2 is going to involve other people using our API, to avoid people reporting issues in the future its worth following the existing specfiications IMO